src/Microsoft.AspNetCore.WebHooks.Receivers/WebHookAttribute.cs

```diff
@@ -0,0 +1,126 @@
+    /// <see cref="WebHookReceiverExistsFilter"/> for the action.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = false, Inherited = true)]
+    public abstract class WebHookAttribute : Attribute, IAllowAnonymous, IFilterFactory
+    {
+        private string _id;
+
+        // TODO: Move attribute constructors' comments, especially the recommended action signatures, to class level.
+        // TODO:  Do the same for all subclasses. As-is important information is hard to find.
+        /// <summary>
+        /// <para>
+        /// Instantiates a new <see cref="WebHookAttribute"/> indicating the associated action is a WebHook
+        /// endpoint for all enabled receivers.
+        /// </para>
+        /// <para>The signature of the action should be:
+        /// <code>
+        /// Task{IActionResult} ActionName(string receiverName, string id, string[] events, TData data)
+        /// </code>
+        /// or the subset of parameters required. <c>TData</c> must be compatible with expected requests.
+        /// </para>
+        /// <para>This constructor should usually be used at most once in a WebHook application.</para>
+        /// <para>
+        /// The default route <see cref="Mvc.Routing.IRouteTemplateProvider.Name"/> is <see langword="null"/>.
+        /// </para>
+        /// </summary>
+        protected WebHookAttribute()
+        {
+        }
+
+        /// <summary>
+        /// <para>
+        /// Instantiates a new <see cref="WebHookAttribute"/> indicating the associated action is a WebHook
+        /// endpoint for the given <paramref name="receiverName"/>.
+        /// </para>
+        /// <para>The signature of the action should be:
+        /// <code>
+        /// Task{IActionResult} ActionName(string id, string[] events, TData data)
+        /// </code>
+        /// or include the subset of parameters required. <c>TData</c> must be compatible with expected requests.
+        /// </para>
+        /// <para>
+        /// This constructor should usually be used at most once per <paramref name="receiverName"/> name in a WebHook
+        /// application.
+        /// </para>
+        /// <para>
+        /// The default route <see cref="Mvc.Routing.IRouteTemplateProvider.Name"/> is <see langword="null"/>.
+        /// </para>
+        /// </summary>
+        /// <param name="receiverName">The name of an available <see cref="IWebHookReceiver"/>.</param>
+        protected WebHookAttribute(string receiverName)
+        {
+            if (string.IsNullOrEmpty(receiverName))
+            {
+                throw new ArgumentException(Resources.General_ArgumentCannotBeNullOrEmpty, nameof(receiverName));
+            }
+
+            ReceiverName = receiverName;
+        }
+
+        /// <summary>
+        /// Gets the name of an available <see cref="IWebHookReceiver"/>.
+        /// </summary>
+        public string ReceiverName { get; }
```

Comment:
>public string ReceiverName { get; } [](start = 8, length = 35)

Is this always required? A subclass that uses the default constructor can't set it

Reply:
Yes, a receiver-specific attribute needs a receiver-specific non-`null` name. `WebHookMetadataProvider` checks this property to make sure it's not `null` except in an attribute that also implements `IWebHookRequestMetadata`. But, there isn't a check to make sure the attribute is `[GeneralWebHook]`.

Easiest thing to do may be to make the default constructor above `internal`. Thoughts?